workbench-core/datasets/src/dataSetService.ts

```diff
@@ -245,16 +245,32 @@ export class DataSetService {
    *
    * @returns an array of DataSet objects.
    */
-  public async listDataSets(authenticatedUser: { id: string; roles: string[] }): Promise<DataSet[]> {
+  public async listDataSets(authenticatedUser: AuthenticatedUser): Promise<DataSet[]> {
     const metadata: Metadata = {
       actor: authenticatedUser,
       action: this.listDataSets.name,
       source: {
         serviceName: DataSetService.name
       }
     };
+
     try {
-      const response = await this._dbProvider.listDataSets();
+      const response: DataSet[] = [];
+      const allDatasets = await this._dbProvider.listDataSets();
+
+      const allDatasetsPermissions = await Promise.all(
+        allDatasets.map((dataset) =>
+          this._getAuthenticatedUserDatasetPermissions(authenticatedUser, dataset.id)
+        )
+      );
+
+      allDatasetsPermissions.forEach((permission, index) => {
+        if (permission.length) {
+          //eslint-disable-next-line security/detect-object-injection
+          response.push(allDatasets[index]);
```

Comment:
Given you are building `allDatasetsPermissions` using potentially several asynchronous calls, are you guaranteed the indexes in `allDatasetsPermissions` will map directly to the same index in `allDataSets`?

Reply:
Yes, since `allDataSetsPermissions` is generated by a map function, the permissions there will correlate to the dataset at the same index.